glib-macros/src/clone.rs

```diff
@@ -0,0 +1,562 @@
+        "default-return" => panic!("`@default-return` should be after `=>`"),
+        "default-panic" => panic!("`@default-panic` should be after `=>`"),
+        k => panic!(
+            "Unknown keyword `{}`, only `weak`, `weak-allow-none` and `strong` are allowed",
+            k,
+        ),
+    };
+    let name = get_full_ident(parts, borrow_kind);
+    let alias = match parts.peek() {
+        Some(TokenTree::Ident(p)) if p.to_string() == "as" => {
+            parts.next();
+            match parts.next() {
+                Some(TokenTree::Ident(i)) => Some(i.to_string()),
+                Some(x) => panic!(
+                    "Expected ident after `as` keyword, found `{}`",
+                    x.to_string()
+                ),
+                None => panic!("Unexpected end after `as` keyword"),
+            }
+        }
+        Some(TokenTree::Ident(p)) => panic!("Unexpected `{}`", p.to_string()),
+        _ => None,
+    };
+    if name == "self" && alias.is_none() {
+        panic!(
+            "Can't use `self` as variable name. Try storing it in a temporary variable or \
+                rename it using `as`."
+        );
+    } else if name.ends_with(".") {
+        panic!("Invalid variable name: `{}`", name);
+    } else if name.contains(".") && alias.is_none() {
+        panic!(
+            "`{}`: Field accesses are not allowed as is, you must rename it!",
+            name
+        );
+    }
+    elements.push(ElemToClone {
+        name,
+        alias,
+        borrow_kind,
+    });
+}
+
+fn group_to_string(g: &Group) -> String {
+    format!(
+        "{}{}{}",
+        match g.delimiter() {
+            Delimiter::Parenthesis => "(",
+            Delimiter::Brace => "{",
+            Delimiter::Bracket => "[",
+            Delimiter::None => "",
+        },
+        tokens_to_string(g.stream().into_iter().peekable()),
+        match g.delimiter() {
+            Delimiter::Parenthesis => ")",
+            Delimiter::Brace => "}",
+            Delimiter::Bracket => "]",
+            Delimiter::None => "",
+        },
+    )
+}
+
+fn get_expr(parts: &mut Peekable<ProcIter>) -> String {
+    let mut ret = String::new();
+    let mut total = 0;
+    match parts.next() {
+        Some(TokenTree::Literal(l)) => ret.push_str(&l.to_string()),
+        Some(TokenTree::Ident(i)) => ret.push_str(&i.to_string()),
+        Some(TokenTree::Punct(p)) => match p.to_string().as_str() {
+            "[" | "{" | "(" => {
+                total += 1;
+            }
+            x => panic!("Unexpected token `{}` after `@default-return`", x),
+        },
+        Some(TokenTree::Group(g)) => return group_to_string(&g),
+        None => panic!("Unexpected end after `@default-return`"),
+    };
+    loop {
+        match parts.peek() {
+            Some(TokenTree::Punct(p)) => {
+                let p_s = p.to_string();
+                if p_s == "{" || p_s == "(" || p_s == "[" {
+                    total += 1;
```

Comment:
For example this kind of parsing should already be taken care of by syn. It parses the token tree into a proper Rust AST. I worry that if we parse Rust code ourselves here that we end up missing various special cases and have problems maintaining this over time (e.g. if the language is extended).